### ISSUE TITLE
range.getImage() returns upside-down image on Excel for Mac

### ISSUE THREAD

**Issue by Andrew-AnalysisPlace** [OfficeDev/office-js#235]
rangeObject.GetImage() works great in Excel for Windows, partially works in Excel Online, and returns an upside-down image in Excel for Mac 16.16.1.
I use range.getImage() to get a picture of exactly what is in a range. This is then used in other Office applications.

- In Excel for Windows (version 1806), it works perfectly as I had hoped. It gets anything put on the range, including cell content, images, shapes, etc. pasted in the range.
- In Excel Online, it simply gets the cell content, including text, borders, background color, and conditional formatting, but not other things pasted on top of the range.
- In Excel for Mac (version 16.15), it only returns a black image (with the correct dimensions).
- In Excel for Mac (version 16.16.1), it returns an upside-down image that does not have the correct colors (blue instead of black). It also does not have shapes/images inserted on top of the range.

## Expected Behavior
Same image appearance in Office for Mac as Windows

## Current Behavior
Office for Mac returns upside-down image

## Steps to Reproduce, or Live Example
Code is here:  https://stackoverflow.com/questions/51679941/office-js-excel-issues-with-range-getimage

## Context
I/customers use range.getImage() to get a picture of exactly what is in a range. This is then used in other Office applications (Word and PowerPoint).

## Your Environment
see above

**Comment by XiangGaoMSFT** [comment]
Thanks for reporting this inconsistent behavior.

I tried [this gist](https://gist.github.com/XiangGaoMSFT/c2f9f8f8c4513d0126d7153f238051b3)  and confirmed below behaviors:
* On Excel Online, shapes and charts are not returned
* On Excel for Mac (16.18), it has the same behavior as the user reported on 16.16.1

Will talk to our PM and respond later.

**Comment by Andrew-AnalysisPlace** [comment]
Tested on Mac 16.19. The issue still occurs (upside-down and incorrect colors). 
My customers use this feature (get image of a range) often on Windows and would like to be able to support it on Mac too. 
Any updates?

**Comment by Andrew-AnalysisPlace** [comment]
Multiple customers still report this issue/limitation on Macs. Any updates?

**Comment by grangeryy** [comment]
Sorry for the late response. I've created a bug to track this issue: Bug 3454384

**Comment by Andrew-AnalysisPlace** [comment]
Every week I still have a couple new Mac customers report this issue. Please provide an update. When should I tell my customers this should be fixed?

**Comment by lumine2008** [comment]
Hi @Andrew-AnalysisPlace , I am very sorry about the experience of this issue, our team has done investigation, and tried to fix the issue, however, the code didn't check-in, because based on team's investigation, this is a not bug level fix, it would lead to other potential issues. so although this is a bug, but we treat it more like adding a new feature. As it take much longer time than we expected. we don't want to create another more serious bug because of fixing this issue. So would you please share us: the MAU of your add-in? what's your scenario of using range.getImage() API?

**Comment by Andrew-AnalysisPlace** [comment]
Thanks for the update @lumine2008 
Our add-in is https://appsource.microsoft.com/en-us/product/office/WA104380955
Customers use it to update content (such as images of ranges) in Word or PowerPoint based on content in Excel. Here are customer scenarios (use cases): https://analysisplace.com/Use-Cases.
For example, in our add-in, if you "Insert Sample Content" in Excel and insert/view the {Chart:Start!rpt_Features_img} item in Word or PowerPoint, it looks like the below on Macs.
![image](https://user-images.githubusercontent.com/28869066/72368160-7e183c80-36cb-11ea-91ea-27e692e75cc5.png)

A possibly related issue (on all platforms): chart.getImage() often/usually crops off the right and/or bottom borders (1 px). This is a more-common customer complaint because most people want to transfer charts and most charts have borders by default. Should I open a new Issue for this? Here is a typical example (bottom border cropped off): 
![image](https://user-images.githubusercontent.com/28869066/72370294-ee28c180-36cf-11ea-8b65-4fb50ada9bfb.png)

**Comment by janisdd** [comment]
I'm also on Excel (16.33) for mac and have the same issue.
The upside down image can be fixed relatively easy inside the add-in with various js image libraries...
For me the main issue is the color problem as it can't be fixed by the add-in itself.

**Comment by ghost** [comment]
Seeing as this issue is still a thing and it looks like it is going to be hard to add this feature to Excel running on Mac, is there any alternative solutions (some npm package for e.g.)? 
Thanks for the hard work you've put into this so far!

**Comment by ghost** [comment]
I was thinking, as an alternative to this function since it is possible to retrieve base64 data about a selection or a worksheet, would it be possible to somehow convert that data to a base64 image?

**Comment by janisdd** [comment]
I'm not sure what function do you mean but I guess it would only return the data (maybe also styles) in that selection/range.
Getting data from a range can also be done with other libraries... but no library (I'm aware of) can create an image from a range/selection like excel would render it.

**Comment by ghost** [comment]
> I'm not sure what function do you mean but I guess it would only return the data (maybe also styles) in that selection/range.
> Getting data from a range can also be done with other libraries... but no library (I'm aware of) can create an image from a range/selection like excel would render it.

I was thinking of this function https://docs.microsoft.com/en-us/javascript/api/excel?view=excel-js-preview#excel-createworkbook-base64-  but you are probably right, I don't think there is any alternative to Excel's getImage().

**Comment by ghost** [comment]
> I'm also on Excel (16.33) for mac and have the same issue.
> The upside down image can be fixed relatively easy inside the add-in with various js image libraries...
> For me the main issue is the color problem as it can't be fixed by the add-in itself.

Would you care to share your code or what libraries you used? For now I am only extracting simple black text and numbers, so I will probably be alright with the current state of getImage().

**Comment by janisdd** [comment]
Sure. I used `jimp` https://github.com/oliver-moran/jimp/tree/master/packages/jimp to rotate the image (e.g. via cdn https://cdn.jsdelivr.net/npm/jimp@0.10.3/browser/lib/jimp.js).

Then do this

```js
//here you set the base64 string
var base64Img = setImg()

//remove the "header stuff"
let plainBase64 = base64Img.substring('data:image/jpeg;base64,'.length)

//this is for a html example
let outImg = document.getElementById('outImg')
let beforeImg = document.getElementById('beforeImg')

//show the normal img
beforeImg.src = base64Img

Jimp.read(_base64ToArrayBuffer(plainBase64)).then(img => {
  img.rotate(180)
  img.getBase64(Jimp.AUTO, (err, res) => {
    //res is your base64 rotated img
    outImg.src = res
  })
})

function _base64ToArrayBuffer(base64) {
    var binary_string = window.atob(base64);
    var len = binary_string.length;
    var bytes = new Uint8Array(len);
    for (var i = 0; i < len; i++) {
        bytes[i] = binary_string.charCodeAt(i);
    }
    return bytes.buffer;
}
```

To test it in a browser you just need to add the following tags to some X.html

```html
<img id="beforeImg" src="" />
<hr />
<img id="outImg" src="" />
```

I only tested this in the browser but it should work in the add-in (I guess).

**Comment by david-ps** [comment]
@ElizabethSamuel-MSFT Any ETA or update on this?

**Comment by ElizabethSamuel-MSFT** [comment]
@david-ps Apologies for my late response. @gingerjia Any update on the backlog item?

Thanks.

**Comment by gingerjia** [comment]
Hi @david-ps , the issue is in still our backlog. May I know the @janisdd's workaround is suitable for you? And also may I understand the impact to your add-in? Then we can priority the work with PM based on it. Thanks very much!

**Comment by Andrew-AnalysisPlace** [comment]
@janisdd's workaround is not suitable in our case because it does not fully correct the image (colors are not close).
Impact to our add-in: thousands of customers use our add-in; 20% are Mac users; approx a third of those would want to use the getRange() to transfer images of ranges to Word/PowerPoint docs (like Windows users do). We do recommend a workaround to our users: paste the range as a "linked picture", then link that shape (but most users would prefer not to use the workaround).

**Comment by gingerjia** [comment]
@Andrew-AnalysisPlace  thanks very much for the quick reply and clear impact info. I have updated the info to our backlog item [BUG 4959708]. Since we did investigation before and found it's not a small effort fix, so during our side try to make progress for it, may still your kindly help to unblock your customer by the workaround as much as possible. We will keep you update if have progress.
Thanks very much again for your help to improve our quality!

**Comment by dingjin-ms** [comment]
Hi, we have a temporary fix for this issue on Mac, and the fix is already available in the latest production version for Mac. Therefore I'm closing this issue for now. Please feel free to re-open the issue if you encounter any other problems. Thank you!
